cli/config.go

```diff
@@ -0,0 +1,173 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cli
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"github.com/sourcenetwork/defradb/logging"
+)
+
+const (
+	configStoreBadger = "badger"
+	configStoreMemory = "memory"
+)
+
+// configPaths are config keys that will be made relative to the rootdir
+var configPaths = []string{
+	"datastore.badger.path",
+	"api.pubkeypath",
+	"api.privkeypath",
+}
+
+// configFlags is a mapping of config keys to cli flags to bind to.
+var configFlags = map[string]string{
+	"log.level":                         "loglevel",
+	"log.output":                        "logoutput",
+	"log.format":                        "logformat",
+	"log.stacktrace":                    "logtrace",
+	"log.nocolor":                       "lognocolor",
+	"api.address":                       "url",
+	"datastore.maxtxnretries":           "max-txn-retries",
+	"datastore.store":                   "store",
+	"datastore.badger.valuelogfilesize": "valuelogfilesize",
+	"net.peers":                         "peers",
+	"net.p2paddresses":                  "p2paddr",
+	"net.p2pdisabled":                   "no-p2p",
+	"api.allowed-origins":               "allowed-origins",
+	"api.pubkeypath":                    "pubkeypath",
+	"api.privkeypath":                   "privkeypath",
+}
+
+// defaultConfig returns a new config with default values.
+func defaultConfig() *viper.Viper {
+	cfg := viper.New()
+
+	cfg.AutomaticEnv()
+	cfg.SetEnvPrefix("DEFRA")
+	cfg.SetEnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_"))
+
+	cfg.SetConfigName("config")
+	cfg.SetConfigType("yaml")
+
+	cfg.SetDefault("datastore.badger.path", "data")
+	cfg.SetDefault("net.pubSubEnabled", true)
+	cfg.SetDefault("net.relay", false)
+	cfg.SetDefault("log.caller", false)
+
+	return cfg
+}
+
+// createConfig writes the default config file if one does not exist.
+func createConfig(rootdir string, flags *pflag.FlagSet) error {
+	cfg := defaultConfig()
+	cfg.AddConfigPath(rootdir)
+
+	if err := bindConfigFlags(cfg, flags); err != nil {
+		return err
+	}
+	// make sure rootdir exists
+	if err := os.MkdirAll(rootdir, 0755); err != nil {
+		return err
+	}
+	err := cfg.SafeWriteConfig()
+	if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok { //nolint:errorlint
+		return nil
+	}
+	return err
+}
+
+// loadConfig returns a new config with values from the config in the given rootdir.
+func loadConfig(rootdir string, flags *pflag.FlagSet) (*viper.Viper, error) {
+	cfg := defaultConfig()
+	cfg.AddConfigPath(rootdir)
+
+	// attempt to read the existing config
+	err := cfg.ReadInConfig()
+	if _, ok := err.(viper.ConfigFileNotFoundError); err != nil && !ok { //nolint:errorlint
+		return nil, err
+	}
+	// bind cli flags to config keys
+	if err := bindConfigFlags(cfg, flags); err != nil {
+		return nil, err
+	}
+
+	// make paths relative to the rootdir
```

Comment:
praise: Thank you for comments like this, it helps a lot when the code itself is non-obvious.